Add docs for Customer Types feature (#1815)
saleor/saleor-docs: docs/developer/users/customer-types.mdx

Reviewer: I think that there is one more behavior worth documenting: customer attribute values are excluded from the translations API - `translations(kind: ATTRIBUTE_VALUE)` skips them. Since product and page attribute values are translatable, it can be assumed that customer values are too. Maybe a short note in this section would do, something like:                                                                            

> Values of customer attributes may contain personal data and are excluded from the translations API - they are not listed as translatable content and cannot be translated

Author: Added.

---
title: Customer Types and Attributes
description: Group customers with customer types and attach structured attribute data to them.
---

**Added in Saleor 3.23.**

Customer types let you segment customers and attach structured data to their accounts. A customer type is a named collection of attributes — the same [attribute system](/developer/attributes/overview.mdx) used by product and page types — so you can model concepts like loyalty tiers, company details for B2B accounts, or onboarding metadata, and have them validated and queryable through the GraphQL API.

> Every customer belongs to exactly one customer type. The type determines which attributes can be set on the customer and which attribute values are visible when reading the customer.

Reviewer: ```suggestion
Every customer belongs to exactly one customer type. The type determines which attributes can be set on the customer and which attribute values are returned when querying the customer.
```

```graphql
mutation {
  customerUpdate(
    id: "VXNlcjoyMw=="
    input: {
      customerType: "Q3VzdG9tZXJUeXBlOjE="
      attributes: [
        { id: "QXR0cmlidXRlOjQ3", dropdown: { value: "Gold" } }
      ]
    }
  ) {
    user {
      id
    }
    errors {
      field
      code
      message
    }
  }
}
```

The attributes you pass must be assigned to the customer type the user ends up with — the type set in the same mutation, or the user's current type when `customerType` is not provided. Passing an attribute outside that type returns a `NOT_FOUND` error.

[`customerBulkUpdate`](/api-reference/users/mutations/customer-bulk-update) accepts the same `customerType` and `attributes` fields, with the same validation applied per customer.

Changing a customer's attribute values triggers the `CUSTOMER_UPDATED` webhook event.

## Reading Attribute Values

Read a customer's type and attribute values through the `User` type:

```graphql
query {
  user(id: "VXNlcjoyMw==") {
    customerType {
      name
    }
    assignedAttributes {
      attribute {
        slug
      }
    }
    assignedAttribute(slug: "loyalty-level") {
      attribute {
        slug
      }
    }
  }
}
```

> `assignedAttributes` returns the attributes of the user's current customer type together with their values. Staff with `MANAGE_USERS` see all of them, while the account owner sees only attributes visible in the storefront.

Reviewer: ```suggestion
`assignedAttributes` returns the attributes of the user's current customer type together with their values. Staff or apps with `MANAGE_USERS` see all of them, while the account owner sees only attributes visible in the storefront.
```

## Filtering Customers

The [`customers`](/api-reference/users/queries/customers) query supports filtering by customer type and by attribute values:

```graphql
query {
  customers(
    first: 10
    where: {
      customerType: { eq: "Q3VzdG9tZXJUeXBlOjE=" }
      attributes: [{ slug: "loyalty-level", value: { slug: { eq: "gold" } } }]
    }
  ) {
    edges {
      node {
        email
      }
    }
  }
}
```

Filtering by the default customer type also matches users that were never explicitly assigned a type.

## Attribute Value Lifecycle

> Attribute values are stored independently of the customer type they were written under. Visibility is evaluated at read time against the user's current type:

Reviewer: ```suggestion
Attribute values are stored independently of the customer type they were written under. Visibility is evaluated at read and filter time against the user's current type:
```
